backup.1
.TH "BACKUP" 1 "2015-01-15" "Robert L Parker rlp1938@gmail.com"


.SH NAME

.P
\fBbackup\fR \- A program to backup the invoking user's home directory by
running \fBrsync\fR.

.SH SYNOPSIS

.P
\fBbackup\fR \- requires no arguments and no options.

.SH DESCRIPTION

.P
\fBbackup\fR \- determines the source and destination paths from
information recorded in configuration files located in
\fI$HOME/.config/backup/\fR. The destination is to be in a removable
device connected via USB or other suitable technology. Generally
speaking \fBbackup\fR is \fBcron\fR safe regardless of whether \fBcron\fR has
an available MTA or not. But see \fBFIRSTRUN\fR below for an exception.

.P
NB this program is intended to be run by \fBcron\fR at frequent intervals.
Accordingly \fBrsync\fR is run without using the \fB\-\-del\fR option.
But it is run a second time using both
\fB--del\fR and \fB--dry-run\fR options and logged separately to the
main log. See \fBFILES\fR below for more detail.

.SH OPTIONS

.TP
 \fB\-h\fR
print help information and exit.

.TP
 \fB\-b\fR
\fIprogram_or_script\fR to run before the actual backup takes place.
Repeat this option for as many programs or scripts that you want to run
before the backup happens. There is no limit, other than available
memory, on the number of programs to run.
.TP
 \fB\-a\fR
\fIprogram_or_script\fR to run after the actual backup finishes.
You may repeat this option as often as needed as above.


.SH FIRSTRUN

.P
There are 3 configuration files that must be installed for \fBbackup\fR
and associated utilities to function. These are \fIbackup.cfg\fR,
\fIexcludes\fR and \fIcruft\fR.
The destination dir for these files is \fI$HOME/.config/backup\fR.
So the expectation is that the first time you run \fBbackup\fR the config
files will be installed and you will edit them to ensure that they meet
your needs. The program expects that the original files will be
installed at \fI/usr/local/share/backup/\fR. If you have used some other
$PREFIX at the configuration stage of the build you will need to copy
these files yourself to the required location. See \fBFILES\fR for a
detailed description of the files that \fBbackup\fR uses and generates.

.SH FILES

.P
\fIbackup.cfg\fR names the backup device eg \fI/dev/sdb/\fR without the
partition number, and the name of the backup directory. You may use more
than one backup set, and may have them with several partitions. The
program will find the named backup dir if the device is mounted, or
record an error if not.

.P
\fIexcludes\fR is to contain a list of what you do not want to be backed
up. The items are to be listed in a way understood by \fBrsync\fR.

.P
\fIcruft\fR More than 90% of backup activity is generated by programs like
browsers writing and deleting cache and config files. This generates a
huge amount of data that may mask events that need to be seen by the
user. \fBdecruft\fR is run using a logfile as it's argument and it lists
the content of that logfile on \fIstdout\fR but it excludes those items that match regular expressions recorded in \fIcruft\fR. These regexes are
to be in \fBegrep\fR format. See \fBEXAMPLES\fR below.

.P
\fI$HOME/lock/backup.lock\fR A zero length file to guard against multiple
versions of \fBbackup\fR running. The file is removed when \fBbackup\fR
completes.

.P
\fI$HOME/log/backup.log\fR The main log file. If viewed when \fBbackup\fR is
running you may see lines beginning with 'deleting'. These are produced
by the second run of \fBrsync\fR and will be removed when \fBbackup\fR
finishes. This log file grows continually until \fBbulogrot\fR is run.

.P
\fI$HOME/log/dryrun.log\fR This file is created new at the end of every
run of \fBbackup\fR. Because lines beginning with 'deleting' will be
generated on every run there is no point in appending this file.

.P
\fI$HOME/log/errors.log\fR This file captures everything written to
\fIstderr\fR. Generally it will record events such as the backup drive
not mounted or that an instance of backup is already running.

.P
\fI$HOME/log/backup.log.0\fR .. \fIbackup.log.7\fR generated by \fBbulogrot\fR
\&. \fIbackup.log.7\fR is discarded on each run of \fBbulogrot\fR.

.P
\fI$HOME/log/errors.log.0\fR .. \fIerrors.log.7\fR generated by \fBbulogrot\fR
\&. \fIerrors.log.7\fR is discarded on each run of \fBbulogrot\fR.

.SH EXAMPLES

.P
decruft ~/log/dryrun.log |grep '^deleting' |less

.P
decruft ~/log/backup.log |less

.P
cleanup

.P
backup

.P
bulogrot

.SH SEE ALSO

.P
\fBbulogrot\fR (1), \fBcleanup\fR (1), \fBdecruft\fR (1)

.\" man code generated by txt2tags 2.6 (http://txt2tags.org)
.\" cmdline: txt2tags -t man backup.t2t


backup.c
/*      backup.c
 *
 *	Copyright 2014/12/23 Bob Parker <rlp1938@gmail.com>
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *	MA 02110-1301, USA.
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <sys/stat.h>
#include <string.h>
#include <dirent.h>
#include <ctype.h>
#include <limits.h>
#include <libgen.h>
#include <errno.h>

#include "fileutil.h"
#include "runutils.h"
#include "backutils.h"
#include "greputils.h"
#include "checkps.h"


// Globals
static char *frfmt =
"This the first run of backup. Backup needs to copy three files, "
"%s, %s and %s from "
"/usr/local/share. The copies will be installed in %s. You should edit "
"these files to ensure they meet your needs. In the configuration file "
"you name the device on which the backup directory resides and also "
"name the actual backup directory. There is no need to name the device "
"number. Backup will find the named backup dir if indeed a backup "
"device is mounted. The minimum content of the exclusions file is "
"the set of files that are altered by the backup program itself. You "
"can add to that anything you don't want to backup. If you used a "
"prefix other than /usr/local when you built the program you will need "
"to copy the above named files into your config directory by hand."
;

char *helpmsg = "\n\tUsage: backup [option]\n"
  "\n\tOptions:\n"
  "\t-h outputs this help message and exits.\n"
  "\t-b program_or_script. Runs the named program before the actual\n"
  "\t   backup starts. The option may be repeated for as many programs\n"
  "\t   are required to run. There is no limit other than available\n"
  "\t   memory. \n"
  "\t-a program_or_script. Runs the named program after the backup\n"
  "\t   completes. The option may be repeated as above.\n"
  ;

void dohelp(int forced);


int main(int argc, char **argv)
{
	char home[NAME_MAX], user[NAME_MAX];
	struct fdata conf, logs;
	char dirname[NAME_MAX], bdev[32], wrkdir[PATH_MAX],
				command[NAME_MAX];
	off_t fsize;
	FILE *fpo, *fpe;
	int opt;

	char **runbefore = malloc(10 * sizeof(char *));
	int beforeruns = 0;
	int beforemax = 10;
	char **runafter = malloc(10 * sizeof(char *));
	int afterruns = 0;
	int aftermax = 10;
	memset(runbefore, 0, beforemax * sizeof(char *));
	memset(runafter, 0, aftermax * sizeof(char *));

	while((opt = getopt(argc, argv, ":hb:a:")) != -1) {
		switch(opt){
		case 'h':
			dohelp(0);
		break;
		case 'b': // run before
		if (beforeruns > beforemax - 1) {
			// always leave a terminating NULL at the end of the list.
			runbefore = realloc(runbefore, (beforemax + 10) *
								sizeof(char *));
			// beforemax is the offset into new runbefore to be zeroed.
			memset(runbefore + beforemax, 0, 10 * sizeof(char *));
			beforemax += 10;
		}
		runbefore[beforeruns] = dostrdup(optarg);
		beforeruns++;
		break;
		case 'a': // run after
		if (afterruns > aftermax - 1) {
			// always leave a terminating NULL at the end of the list.
			runafter = realloc(runafter, (aftermax + 10) *
								sizeof(char *));
			// aftermax is the offset into new runafter to be zeroed.
			memset(runafter + aftermax, 0, 10 * sizeof(char *));
			aftermax += 10;
		}
		runafter[afterruns] = dostrdup(optarg);
		afterruns++;
		break;
		case ':':
			fprintf(stderr, "Option %c requires an argument\n",optopt);
			dohelp(1);
		break;
		case '?':
			fprintf(stderr, "Illegal option: %c\n",optopt);
			dohelp(1);
		break;
		} //switch()
	}//while()

	strcpy(progname, basename(argv[0]));
	/* The POS below successfully stops gcc bitching but then gives
	 * cppcheck something to barf about. I'll leave it like that as gcc
	 * is run often but cppcheck much less.
	*/
	if(argc) argc = argc;

	dogetenv("HOME", home);
	// prepare home to suit rsync.
	if (home[strlen(home) -1 ] != '/') strcat(home, "/");

	// setup my file paths
	sprintf(wrkdir, "%s.config/backup/backup.cfg", home);
	char *cfgfile = dostrdup(wrkdir);	// config file.
	sprintf(wrkdir, "%slock/backup.lock", home);
	char *lockfile = dostrdup(wrkdir);	// lock file
	sprintf(wrkdir, "%slog/backup.log", home);
	char *logfile = dostrdup(wrkdir);	// log file
	sprintf(wrkdir, "%slog/dryrun.log", home);
	char *cruftfile = dostrdup(wrkdir);	// cruft
	sprintf(wrkdir, "%slog", home);
	char *logdir = dostrdup(wrkdir);	// log dir, to mkdir if needed.
	sprintf(wrkdir, "%slock", home);
	char *lockdir = dostrdup(wrkdir);	// lock dir, to mkdir if needed.
	sprintf(wrkdir, "%s.config/backup/", home);
	char *cfgdir = dostrdup(wrkdir);	// config dir.
	sprintf(wrkdir, "%sexcludes", cfgdir);
	char *exclfile = dostrdup(wrkdir);	// excludes file
	sprintf(wrkdir, "%slog/trace.txt", home);
	char *tracefile = dostrdup(wrkdir);	// trace wtf
	sprintf(wrkdir, "%slog/bulog.bak", home);
	char *bulog = dostrdup(wrkdir);	// safety copy of the log.
	sprintf(wrkdir, "%slog/errors.log", home);
	char *errlog = dostrdup(wrkdir);	// log errors.
	sprintf(wrkdir, "%s.config/backup/cruft", home);
	char *cruftregex = dostrdup(wrkdir);	// describe cruft by regex.

	dogetenv("LOGNAME", user);


	// Is this the first run of backup?
	if (filexists(cfgfile, &fsize) == -1 ||
		filexists(exclfile, &fsize) == -1 ||
		filexists(cruftregex, &fsize) == -1 ) {
		// the first run text
		sprintf(wrkdir, frfmt, "excludes", "backup.cfg", "cruft",
					cfgdir);
		char *txt = dostrdup(wrkdir);
		firstrun(txt, "/usr/local/share/backup/", cfgdir, "backup.cfg",
					"excludes", "cruft", NULL);
		free(txt);
		fputs("\nPlease edit these files and run backup again.\n",
				stdout);
		// do the log and lock dirs exist in $HOME? If not make 'em.
		if (direxists(logdir) == -1) {
			// make it then.
			sprintf(command, "mkdir %s", logdir);
			dosystem(command);
		}
		if (direxists(lockdir) == -1) {
			// make it then.
			sprintf(command, "mkdir %s", lockdir);
			dosystem(command);
		}
		goto finis;
	}

	// redirect stdout, stderr
	fpo = dofreopen(logfile, "a", stdout);
	fpe = dofreopen(errlog, "a", stderr);

	// nice() this thing, I don't want the slightest delay from it.
	errno = 0;
	if (nice(19) == -1) {
		// -1 may be a legitimate return value.
		if(errno){
			perror("nice()");
			exit(EXIT_FAILURE);
		}
	}
	conf = readfile(cfgfile, 0, 1);	// config stuff.
	getconfig(conf.from, conf.to, bdev, dirname);
	free(conf.from);

	if (getbupath("/etc/mtab", bdev, dirname, wrkdir) == -1) {
		logthis(progname, "Backup drive is not mounted. Will quit.\n",
				fpe);
		exit(EXIT_FAILURE);
	}

	if (wrkdir[strlen(wrkdir) -1 ] != '/') strcat(wrkdir, "/");
	char *bupath = strdup(wrkdir);

	// Where am I?
	if (!getcwd(wrkdir, PATH_MAX)) {
		perror("getcwd()");
		exit(EXIT_FAILURE);
	}
	if (wrkdir[strlen(wrkdir) -1 ] != '/') strcat(wrkdir, "/");
	if (strcmp(wrkdir, home) != 0) {
		if (chdir(home) == -1) {
			perror(home);
			exit(EXIT_FAILURE);
		}
	}

	// Is an earlier instance of backup running?
	if (filexists(lockfile, &fsize) == 0) {
		// rsync may have crashed us out on an earlier run.
		fclose(fpo);	// checkps needs stdout
		if (checkps() == 0) {
			// found something.
			logthis(progname, "An earlier instance is running."
					"Will quit.", stderr);
			goto finis;
		} else {
			logthis(progname, "An earlier instance had crashed."
					"Will continue.", stderr);
		}
		fpo = dofreopen(logfile, "a", stdout);
	} else {
		// Install the lock file
		FILE *fpl = dofopen(lockfile, "w");
		fclose(fpl);
	}
	sync();

	// process the run before programs/scripts
	int i = 0;
	while(runbefore[i]) {
		char buf[PATH_MAX];
		(void)realpath(runbefore[i], buf);
		dosystem(buf);
		i++;
	}

	// record beginning date
	logthis(progname, "Begin backup.", stdout);
	logthis("Backup destination:", bupath, stdout);
	fclose(fpo);	// Force the writes in correct order
	fpo = dofreopen(logfile, "a", stdout);
	sync();

	// formulate the first command to rsync
	/* This first command does not use the --del option because I do
	 * not want to run the risk of propagating a 'hole', ie missing
	 * file or dir through the backup set.
	* */
	sprintf(command, "/usr/bin/rsync -av --links --hard-links "
	"--exclude-from=%s %s %s", exclfile, home, bupath);
	dosystem(command);

	fclose(fpo);	// Force the writes in correct order
	fpo = dofreopen(logfile, "a", stdout);
	sync();

	// put date finished and a ruleoff.
	logthis(progname, "Ended backup.", stdout);
	fprintf(stdout, "------\n\n");

	fclose(fpo);	// Force the writes in correct order
	fpo = dofreopen(logfile, "a", stdout);
	sync();

	// want to split the dry run stuff out of backup.log to record it
	// separately. So put xml markers around that stuff.
	fprintf(stdout, "%s\n", "<dryrun>");
	logthis(progname, "Begin dry run.", stdout);

	fclose(fpo);	// Force the writes in correct order
	fpo = dofreopen(logfile, "a", stdout);
	sync();

	// formulate the second command to rsync
	/* This second version uses the --del option but also uses
	 * --dry-run and directs the output to $HOME/log/dryrun.log for
	 * manual review of what would have been deleted. Unfortunately
	 * this file is overloaded with browser crap. Run decruft to strip
	 * that out and then manually review what is left before running
	 * cleanup.
	 * */
	sprintf(command, "/usr/bin/rsync -av --dry-run --links --hard-links"
			" --del --exclude-from=%s %s %s", exclfile, home, bupath);
	dosystem(command);

	fclose(fpo);	// Force the writes in correct order
	fpo = dofreopen(logfile, "a", stdout);
	sync();

	// put date finished and a ruleoff.
	logthis(progname, "Ended dry run.", stdout);
	fprintf(stdout, "------\n\n");
	fprintf(stdout, "</dryrun>\n");
	fclose(fpo);	// closes the logfile.
	sync();

	// now split the log
	logs = readfile(logfile, 0, 1);
	if (logs.from) {
		char *lgfr, *lgto, *drfr, *drto;
		dorename(logfile, bulog);	// backup copy of log
		lgfr = logs.from;
		lgto = memmem(lgfr, logs.to - lgfr, "<dryrun>",
						strlen("<dryrun>"));
		if (!lgto) {
			dorename(bulog, logfile);	// keep existing log
			exit(EXIT_FAILURE);
		}
		writefile(logfile, lgfr, lgto);	// new copy of logfile.
		drfr = lgto + strlen("<dryrun>") + 1;
		drto = memmem(drfr, logs.to - drfr, "</dryrun>",
						strlen("</dryrun>"));
		writefile(cruftfile, drfr, drto);	// new copy of cruftfile.
		/* Ok, we made two out of the one. Using the pointers as I have
		 * means that there will be no xml tags in either file. They
		 * are not needed. Apart from that the entire original log is
		 * a concatenation of the two subfiles. I can now safely rm the
		 * original log. NB I write a new copy of the cruft file as all
		 * deleting notifications written to it will repeat until such
		 * time as the cleanup program is run.
		*/
		dounlink(bulog);	// done with it.
	}

	// process the run after programs/scripts
	i = 0;
	while(runafter[i]) {
		char buf[PATH_MAX];
		(void)realpath(runafter[i], buf);
		dosystem(buf);
		i++;
	}

	// clear the lock
	dounlink(lockfile);
	free(logs.from);
	fclose(fpe);	// the error log.
	free(bupath);

finis:
	// free my heap strings
	free(cruftregex);
	free(errlog);
	free(bulog);
	free(tracefile);
	free(cfgdir);
	free(exclfile);
	free(lockdir);
	free(logdir);
	free(cruftfile);
	free(logfile);
	free(lockfile);
	free(cfgfile);

	return 0;
} // main()

void dohelp(int forced)
{
  fputs(helpmsg, stderr);
  exit(forced);
}
